### PR TITLE
[TT-13139] Request times out in some cases when sending input via http inputs

### PR DESCRIPTION
**Pull request by buraksezer**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-13139" title="TT-13139" target="_blank">TT-13139</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Request times out in some cases when sending input via http inputs</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

Cherry-picked stream caching feature from this branch: https://github.com/TykTechnologies/tyk/pull/6538

Two new integration tests have been added to test `input http -> output http` scenario. See this issue for the details: https://tyktech.atlassian.net/browse/TT-13139

Closing the previous one: https://github.com/TykTechnologies/tyk/pull/6592

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented stream caching and garbage collection in the `StreamingMiddleware` to manage inactive streams and improve performance.
- Added new fields and methods to handle stream activity and caching efficiently.
- Introduced a garbage collection routine to periodically clean up inactive stream managers.
- Added integration tests for single and multiple client scenarios, focusing on HTTP server input and WebSocket output.
- Verified message distribution and handling in the new tests to ensure correct functionality.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_streaming.go</strong><dd><code>Implement stream caching and garbage collection in StreamingMiddleware</code></dd></summary>
<hr>

gateway/mw_streaming.go

<li>Introduced stream caching and garbage collection for inactive streams.<br> <li> Added new fields to manage stream activity and cache.<br> <li> Implemented a garbage collection routine for stream managers.<br> <li> Updated stream manager creation to utilize caching.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6601/files#diff-6f565750150d990575c808f1ca8f38483160dc6edf05f1534cd0bedb27c2e6c8">+98/-20</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_streaming_test.go</strong><dd><code>Add integration tests for HTTP server streaming scenarios</code></dd></summary>
<hr>

gateway/mw_streaming_test.go

<li>Added tests for single and multiple client streaming scenarios.<br> <li> Implemented test for HTTP server input and WebSocket output.<br> <li> Verified message distribution and handling in tests.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6601/files#diff-a0d1bd0196a741537a3c850e340225c8993e49d709c838af0f1b48b9893af1da">+137/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by buger** [comment]
<p>I'm a bot and I 👍 this PR title. 🤖</p>

File: gateway/mw_streaming.go [hunk 1]
```diff
@@ -150,20 +190,56 @@ func (s *StreamingMiddleware) Init() {
...
 }
 
 func (s *StreamingMiddleware) createStreamManager(r *http.Request) *StreamManager {
-	newStreamManager := &StreamManager{
-		muxer:  mux.NewRouter(),
-		mw:     s,
-		dryRun: r == nil,
+	streamsConfig := s.getStreamsConfig(r)
+	configJSON, _ := json.Marshal(streamsConfig)
+	cacheKey := fmt.Sprintf("%x", md5.Sum(configJSON))
```
Comment (buger) [inline gateway/mw_streaming.go:213]: 100% should be replaced by smth like sha256 or similar. md5 is too weak (I know it was part of orig implementation, but still..)

Comment (buraksezer) [reply gateway/mw_streaming.go]: Replaced by sha256.

**Review by buger — CHANGES_REQUESTED** [review]
I do not think lastActivity doing its job now. 
The actual activity happens inside the f(w, r), and if there is long websocket connection, which last more then 10 minutes, it will timeout, even if it were active. 

If there is current "f(w, r)" running for given consumer group, it should be counted as active. 

```
h.sm.lastActivity.Store(time.Now())
f(w, r)
h.sm.lastActivity.Store(time.Now())
```

**Comment by buraksezer** [comment]
> I do not think lastActivity doing its job now. The actual activity happens inside the f(w, r), and if there is long websocket connection, which last more then 10 minutes, it will timeout, even if it were active.
> 
> If there is current "f(w, r)" running for given consumer group, it should be counted as active.
> 
> ```
> h.sm.lastActivity.Store(time.Now())
> f(w, r)
> h.sm.lastActivity.Store(time.Now())
> ```

@buger I think wrapping `http.ResponseWriter` and updating `lastActivity` in the `Write` method could be an option.

**Comment by buger** [comment]
@buraksezer 

As alternative, you can just maintain counter of active connections, for each stream manager. 
Literally:

```
h.sm.Inc()
f(w, r)
h.sm.Dec()
```

And in GC just check if counter non 0.

**Comment by buraksezer** [comment]
> @buraksezer
> 
> As alternative, you can just maintain counter of active connections, for each stream manager. Literally:
> 
> ```
> h.sm.Inc()
> f(w, r)
> h.sm.Dec()
> ```
> 
> And in GC just check if counter non 0.

This might be the only option because the underlying TCP connection is hijacked for websocket traffic and this invalidates my solution.

**Review by buger — COMMENTED** [review]
```
// Unload closes and remove active streams
func (s *StreamingMiddleware) Unload() {
	s.Logger().Debugf("Unloading streaming middleware %s", s.Spec.Name)
	totalStreams := 0
	s.streamManagers.Range(func(_, value interface{}) bool {
		manager, ok := value.(*StreamManager)
		if !ok {
			return true
		}
		manager.streams.Range(func(_, _ interface{}) bool {
			totalStreams++
			return true
		})
		return true
	})
	globalStreamCounter.Add(-int64(totalStreams))
```

Do we really need separate streamManagers sync.Map here, since we already doing similar with streamManagersCache?

**Comment by buraksezer** [comment]
> ```
> // Unload closes and remove active streams
> func (s *StreamingMiddleware) Unload() {
> 	s.Logger().Debugf("Unloading streaming middleware %s", s.Spec.Name)
> 	totalStreams := 0
> 	s.streamManagers.Range(func(_, value interface{}) bool {
> 		manager, ok := value.(*StreamManager)
> 		if !ok {
> 			return true
> 		}
> 		manager.streams.Range(func(_, _ interface{}) bool {
> 			totalStreams++
> 			return true
> 		})
> 		return true
> 	})
> 	globalStreamCounter.Add(-int64(totalStreams))
> ```
> 
> Do we really need separate streamManagers sync.Map here, since we already doing similar with streamManagersCache?

I think we do not need `streamManagers` map. It's only used to track `globalStreamCounter`. We can use `streamManagersCache` for this.

**Review by pvormste — APPROVED** [review]

**Pull request merged**